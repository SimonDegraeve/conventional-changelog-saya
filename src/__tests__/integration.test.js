/**
 *
 */
import { dir as createDirectory } from 'tmp';
import exec from 'execa';
import writePkg from 'write-pkg';
import { resolve as resolvePath, basename } from 'path';


/**
 *
 */
jasmine.DEFAULT_TIMEOUT_INTERVAL = 20000; // eslint-disable-line no-undef


/**
 *
 */
const createCommitCommand = command => (message) => {
  const messages = Array.isArray(message) ? message : [message];
  const args = messages.reduce((result, value) => ([...result, '--message', value]), []);
  return command('git', ['commit', '--allow-empty', ...args]);
};


/**
 *
 */
let fakeHash = 1000000;
const fakeHashList = {};
const replaceHashes = changelog => changelog.replace(/[0-9a-f]{7}/g, hash => (
  fakeHashList[hash] ? fakeHashList[hash] : (fakeHashList[hash] = (fakeHash += 1))
));


/**
 *
 */
const getChangelog = (command, contextOpts) => () => {
  const code = `
    var changelog = require(
     '${resolvePath(__dirname, '../../node_modules/conventional-changelog-core')}'
    );
    var through = require(
     '${resolvePath(__dirname, '../../node_modules/through2')}'
    );
    var mockDate = require(
      '${resolvePath(__dirname, '../../node_modules/mockdate')}'
    );
    var preset = require(
     '${resolvePath(__dirname, '../preset')}'
    );

    mockDate.set(0); // 1970-01-01

    changelog({ config: preset, releaseCount: 0, outputUnreleased: true }, ${JSON.stringify(contextOpts)})
     .on('error', error =>
       console.error(error)
     )
     .pipe(through((chunk, encoding, callback) => {
       console.log(chunk.toString());
       callback(null, chunk)
     }));
 `;
  return command('babel-node', ['--eval', code]).then(({ stdout }) => replaceHashes(stdout));
};


/**
 *
 */
const createRepository = ({ withPkg = false, contextOpts = {} } = {}) =>
  new Promise((resolve, reject) => {
    createDirectory({ unsafeCleanup: true }, async (error, path) => {
      if (error) {
        return reject(error);
      }

      const command = (...args) => exec(...args, { cwd: path, env: process.env });

      await command('git', ['config', '--global', 'user.email', 'user@email.com']);
      await command('git', ['config', '--global', 'user.name', 'user']);
      await command('git', ['init']);

      if (withPkg) {
        await writePkg(path, {
          name: 'my-module',
          version: '1.0.0',
          repository: 'github:my-user/my-module',
        });
      }

      return resolve({
        path,
        command,
        name: basename(path),
        commit: createCommitCommand(command),
        changelog: getChangelog(command, contextOpts),
      });
    });
  });


/**
 *
 */
describe('preset integration', () => {
  it('works with no package.json', async () => {
    const { commit, changelog } = await createRepository();
    await commit('extra: Initialize commit');
    const result = await changelog();
    expect(result).toMatchSnapshot();
  });

  it('works with a package.json', async () => {
    const { commit, changelog } = await createRepository({ withPkg: true });
    await commit('extra: Initialize commit');
    const result = await changelog();
    expect(result).toMatchSnapshot();
  });

  it('outputs only commit with a valid type', async () => {
    const { commit, changelog } = await createRepository({ withPkg: true });
    await commit('extra: Initialize commit');
    await commit('Random commit');
    await commit('add: Add feature 1');
    await commit('fix: Fix bug 1');
    const result = await changelog();
    expect(result).toMatchSnapshot();
  });

  it('outputs breaking and deprecating changes', async () => {
    const { commit, changelog } = await createRepository({ withPkg: true });
    await commit(['add: Add feature 1', 'BREAKING CHANGE: This is a breaking change 1']);
    await commit(['add: Add feature 2', 'DEPRECATING CHANGE: This is a deprecating change 1']);
    await commit(['add: Add feature 3', '-breaking-change-']);
    await commit(['add: Add feature 4', '-deprecating-change-']);
    const result = await changelog();
    expect(result).toMatchSnapshot();
  });

  it('outputs linked references', async () => {
    const { commit, changelog } = await createRepository({ withPkg: true });
    await commit(['fix: Fix bug #1']);
    const result = await changelog();
    expect(result).toMatchSnapshot();
  });

  it('outputs comparison links', async () => {
    const { commit, changelog } = await createRepository({ withPkg: true, contextOpts: { previousTag: '0.0.5' } });
    await commit('add: Add feature 1');
    const result = await changelog();
    expect(result).toMatchSnapshot();
  });
});
