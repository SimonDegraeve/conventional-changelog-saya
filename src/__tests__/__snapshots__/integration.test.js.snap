exports[`preset integration outputs breaking and deprecating changes 1`] = `
"<a name=\"1.0.0\"></a>
## 1.0.0
> 1970-01-01

* [1000003](https://github.com/my-user/my-module/commit/1000003) **Feature/Breaking:** Add feature 1
* [1000004](https://github.com/my-user/my-module/commit/1000004) **Feature/Deprecating:** Add feature 2
* [1000005](https://github.com/my-user/my-module/commit/1000005) **Feature/Breaking:** Add feature 3
* [1000006](https://github.com/my-user/my-module/commit/1000006) **Feature/Deprecating:** Add feature 4

### Deprecating Changes

* This is a deprecating change 1

### Breaking Changes

* This is a breaking change 1

"
`;

exports[`preset integration outputs comparison links 1`] = `
"<a name=\"1.0.0\"></a>
## [1.0.0](https://github.com/my-user/my-module/compare/0.0.5...v1.0.0)
> 1970-01-01

* [1000008](https://github.com/my-user/my-module/commit/1000008) **Feature:** Add feature 1

"
`;

exports[`preset integration outputs linked references 1`] = `
"<a name=\"1.0.0\"></a>
## 1.0.0
> 1970-01-01

* [1000007](https://github.com/my-user/my-module/commit/1000007) **Bugfix:** Fix bug [#1](https://github.com/my-user/my-module/issues/1)

"
`;

exports[`preset integration outputs only commit with a valid type 1`] = `
"<a name=\"1.0.0\"></a>
## 1.0.0
> 1970-01-01

* [1000001](https://github.com/my-user/my-module/commit/1000001) **Feature:** Add feature 1
* [1000002](https://github.com/my-user/my-module/commit/1000002) **Bugfix:** Fix bug 1

"
`;

exports[`preset integration works with a package.json 1`] = `
"<a name=\"1.0.0\"></a>
## 1.0.0
> 1970-01-01


"
`;

exports[`preset integration works with no package.json 1`] = `
"<a name=\"Unreleased\"></a>
## Unreleased
> 1970-01-01


"
`;
